### ISSUE TITLE
What do Queue wait time and took elapsed time for execution Warnings (in silo logs) indicate ?

### ISSUE THREAD

**Issue by angshuman-agarwal** [dotnet/orleans#1330]
Hi,
I am just subscribing to a TICK DB and setting the object onto the Grain (I am using my own Persistent Stream Provider): 

```
            IPrice record = new Price(row);
            var key = GetKey(row);
            var priceGrain = GrainClient.GrainFactory.GetGrain<IPriceGrain>(key);

            // this will stream the prices internally
            try
            {
                // awaiting this call slows things down
               await priceGrain.SetPrice(record);
            }
            catch (Exception ex)
            {
                Console.WriteLine("!!!!!!!!!!!!!!!!!!!!!!!!! " + ex);
            }
```

But, I randomly get these warnings : _Any idea, how to avoid this ?_ I want to know if I am doing something wrong in my code ? 

```
[2016-01-25 15:27:07.469 GMT     7	WARNING	101201	Runtime.Scheduler.WorkerPoolThread/0	127.0.0.1:11111]	Queue wait time of 00:00:14.1588018 for Item WorkItemGroup:Name=[Activation: S127.0.0.1:11111:191431207*grn/4D0D8712/00000000+ISF.MI@24fcc993 
```

```
[2016-01-25 15:30:28.678 GMT    10	WARNING	101215	Scheduler.WorkItemGroup	127.0.0.1:11111]	Task [Id=11508311, Status=RanToCompletion] in WorkGroup [Activation: S127.0.0.1:11111:191431207*grn/4D0D8712/00000000+IDWR.L@c8b6e880 #GrainType=MyNamespace.GrainCollection.PriceGrain Placement=RandomPlacement State=Valid] took elapsed time 0:00:00.9799022 for execution, which is longer than 00:00:00.2000000. Running on thread <Runtime.Scheduler.WorkerPoolThread/3, ManagedThreadId=10, Executing Task Id=11508311 Status=RanToCompletion for 00:00:00.9790000 on WorkItem=WorkItemGroup:Name=[Activation: S127.0.0.1:11111:191431207*grn/4D0D8712/00000000+IDWR.L@c8b6e880 #GrainType=MyNamespace.GrainCollection.PriceGrain Placement=RandomPlacement],WorkGroupStatus=Running Executing for 00:00:00.9790000. .>	
```

**Comment by sergeybykov** [comment]
1. The second warning indicates that a method of `PriceGrain` blocked the grain thread for almost a second. This is not supposed to happen if all IO operations are preformed via async APIs.

2. The first one tells that a request was queued up for over 14 seconds before it could start executing, likely because all grain threads were blocked by 1. That the reason blocking calls are not allowed within grain logic - they prevent execution of other requests while the blocked thread has nothing to do other than to wait for an operation (usually IO) to complete.

**Comment by angshuman-agarwal** [comment]
Thanks @sergeybykov 
> The second warning indicates that a method of PriceGrain blocked the grain thread for almost a second. This is not supposed to happen if all IO operations are preformed via async APIs.

I am doing this inside PriceGrain (SetPrice method, which gets fired from the callback as shown above in example). Is it the Console.Writeline call  which might have caused it ? Other than that I cannot think of anything ? I did Console print to see what I am getting in real time so that I can compare with the GUI client else it;s very hard to debug.

```
        public async Task SetPrice(IPrice price)
        {
#if DEBUG
            Console.WriteLine(price.Symbol + ">>>>>" + price);
#endif
            _price = price; // set the price
            await _producer.OnNextAsync(price);   // stream it
```

**Comment by sergeybykov** [comment]
Are there other methods on `PriceGrain` that may be blocking?

Also, is this running with `ServerGC`? Is the silo under a memory pressure with expensive Gen2 collections?

If you can share a full silo log, that may shed additional light on what's going on there.

**Comment by angshuman-agarwal** [comment]
>Are there other methods on PriceGrain that may be blocking?


No. Apart from SetPrice I do this in ActivateAsync
```
        public override Task OnActivateAsync()
        {
            IStreamProvider streamProvider = base.GetStreamProvider(ProviderToUse);
            _streamId = Guid.NewGuid();
            _producer = streamProvider.GetStream<IPrice>(_streamId, StreamNamespace);
            return base.OnActivateAsync();
        }
```


>Also, is this running with ServerGC?


Yes. Pls see the SILO app.config (default from Orleans VS template)

```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <runtime>
    <gcServer enabled="true"/>
    <gcConcurrent enabled="false"/>
  </runtime>
  <startup>
    <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.6"/>
  </startup>
</configuration>
```

>Is the silo under a memory pressure with expensive Gen2 collections?

I do not think so. I am just storing last 500 secs of Price in Persistent Stream & then purging it (like a RollingWindow implementation). And as soon as the price arrive I do "SetPrice()" - _set & steam it_.


>If you can share a full silo log, that may shed additional light on what's going on there.


Full SILO log is attached. Thanks for your time !!

[Machine1-2016-01-25-15.20.04.937Z.zip](https://github.com/dotnet/orleans/files/104079/Machine1-2016-01-25-15.20.04.937Z.zip)

**Comment by gabikliot** [comment]
What is the negative impact that you see in your app? If all works fine, than you might be fine. 

The warnings indicate that the system is working pretty hard, most probably since you created 100 queues with 100 agents all working hard to poll the queue (even if the queue is empty). The system is at 50-60 CPU utilization, which is not huge but still respectable, and with 4 cores, you are expected to see some pressure. With all that, if the application is working just fine, no indication of any problem and only some occasional warning msg in the log saying that the system is dealing with pressure, I would say that it looks pretty good to me!

I would start testing with real prod setup (not `MembershipTableGrain`), more than one silo, real expected workload.

**Comment by sergeybykov** [comment]
To piggyback on what @gabikliot already said. With 100 pulling agent system targets actively competing for the CPU, you seem to be getting quite a bit of contention on the CPU, which may explain occasional slow execution of grain methods. I found ~25 instances of 101215/601215 within the 50 minutes of the log.

**Comment by angshuman-agarwal** [comment]
@sergeybykov @gabikliot - Thanks for your comments. Helped a lot in understanding.

The delay was also due to Console Window I noticed. If you hold the scroll bar on Console for long, that message of timeout will get printed. 

I did the set up now with _4 Queues i.e. equal to number of cores_ 

> I would start testing with real prod setup (not MembershipTableGrain)

Used the SQL SERVER tables. Could see the Orleans server entry _(InstanceName field was populated with the node name correctly)_

> more than one silo 

I tried this set up too but not successful.
**Please see separate issue #1336 where grain activation got massively slowed down with 2 SILOS running on 2 different machines**

All the above set up was on one machine. It ran fine. I could see the details being written in SQL SERVER but could not make much sense out of each column (need to read through what all that data means)

![image](https://cloud.githubusercontent.com/assets/1796126/12589504/0a126a22-c456-11e5-801e-bb2cc6fc4d0f.png)

**Comment by veikkoeeva** [comment]
@angshuman-agarwal Not to hijack this thread, but I believe the statistics table should be refactored. Maybe I should write an official issue of it and refer to it (unless someone else starts it). :)

**Comment by gabikliot** [comment]
@angshuman-agarwal , in your perf. setup, please run all silos and client and SQL on different machines (so you need at least 4 machines if you are testing with 2 silos). Otherwise, the results are hard to compare.
You don't really need to use OrleansStatisticstTable, while MembershipTable and SiloMetrics can be useful for you.

**Comment by angshuman-agarwal** [comment]
@gabikliot : Did this set up today on 2 machines and so far looking good after continuous running for 4.5 hours using Prod data (not full load yet - need to write further code to enable the entire flood gate).

* Publisher + SILO - _Machine1_
* SILO + SQL SERVER + REDIS (for persistent stream provider) - _Machine2_

**Comment by gabikliot** [comment]
:+1: 
Even better would be to run the publisher, and SQL and REDIS each on its own machine. 5 machines total.
I am definitely not an IT expert, but I don't think it is considered a good IT practice to run the database on the same box with the application servers. Unless you are doing small apps, like "mini clusters" with all 5 services on the same box. But then don't expect a super good perf.

**Comment by gabikliot** [comment]
@angshuman-agarwal , can that be closed, or are you still investigating?
